completion returns no results with pattern matching

snjeza opened the pull request:
## What it does
Fixes #2106, https://bugs.eclipse.org/bugs/show_bug.cgi?id=568934#c8 and https://bugs.eclipse.org/bugs/show_bug.cgi?id=568934#c6

## How to test
try CA at |
```
import java.util.ArrayList;
import java.util.List;

public class SimpleJavaClass {
  public static void main(String[] args) {
    Object unknown = new ArrayList<>();

    if (unknown instanceof List things) {
      things.is| // try CA here
      Object nothing = null;
      things.is|
    }
  }
}
```
Without the patch:
![patternwithout](https://github.com/eclipse-jdt/eclipse.jdt.core/assets/388609/120331a5-2f1c-4ad7-b6fd-6db850458dd2)

With the patch:
![pattern](https://github.com/eclipse-jdt/eclipse.jdt.core/assets/388609/665b97dd-8186-409b-9b1e-d746bdccbbb5)

I will add a test.

snjeza commented:
@srikanth-sankaran @iloveeclipse I have updated the PR.

> the fix proposed in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2107 is suspect. You should never have to fiddle with ExtraCompilerModifiers.AccOutOfFlowScope bits.

I left this part out. The PR doesn't touch the bits field.

> See also https://bugs.eclipse.org/bugs/show_bug.cgi?id=568934

It also fixes https://bugs.eclipse.org/bugs/show_bug.cgi?id=568934#c8 and https://bugs.eclipse.org/bugs/show_bug.cgi?id=568934#c6

If you agree, I can add a test.

srikanth-sankaran commented:
No, this fix is not correct. It is kludgy and may work but I can't predict what other problems it may cause down the road. 

Basically you are pairing a `LocalVariableBinding` where a TypeBinding would be the appropriate one. That is recipe for CCE. Even if it doesn't trigger CCE in testing, do we want to resort to such hacky fixes is the question.

The real problem as I outlined on the bug earlier is due to wrong parse tree recovery. Can you check why the scanner doesn't declare EOF upon seeing the completion at things.is| and why it returns the token Object at all ? 

As you can see the two statements get jammed into one 

things.is| Object 

so it appears as though you are completing on a nested type, while you are actually completing on a QNR.

snjeza commented:
> and the parse tree recovered in the non-working case is:

I have tried it and I get:

```
import java.util.ArrayList;
import java.util.List;
public class SimpleJavaClass {
  public SimpleJavaClass() {
  }
  public static void main(String[] args) {
    Object unknown;
    {
      <CompleteOnType:things.is> Object;
    }
  }
}
```
There isn't `if ((unknown instanceof List things))`

> Completion is being attempted on a Name in both instances. Why do we think it is being attempted on a Type ? I can help with reviewing an alternate fix. Thanks
> The real problem as I outlined on the bug earlier is due to wrong parse tree recovery. Can you check why the scanner doesn't declare EOF upon seeing the completion at things.is| and why it returns the token Object at all ?
>     The parser returns CompletionOnQualifiedNameReference in line 10 (it works) and CompletionOnQualifiedTypeReference in line 8 (it doesn't work)
> Right, that is the bug that needs fixing.

I have checked the following code:

```
import java.util.ArrayList;
import java.util.List;
public class SimpleJavaClass2 {
  public static void main(String[] args) {
    Object unknown = new ArrayList<>();
    List list = new ArrayList<>();
    if (unknown instanceof List things) {
      list.is|
      Object nothing = null;
      list.is
    }
    
  }
}
```
the parse tree recovered is
```
import java.util.ArrayList;
import java.util.List;
public class SimpleJavaClass2 {
  public SimpleJavaClass2() {
  }
  public static void main(String[] args) {
    Object unknown;
    List list;
    {
      <CompleteOnType:list.is> Object;
    }
  }
}
```
The parser also returns CompletionOnQualifiedTypeReference. It works because the parser finds the `list` variable.

Based on it and the fix for https://bugs.eclipse.org/bugs/show_bug.cgi?id=287939 I have created a new PR

> Basically you are pairing a LocalVariableBinding where a TypeBinding would be the appropriate one. That is recipe for CCE. Even if it doesn't trigger CCE in testing, do we want to resort to such hacky fixes is the question.

This part is removed.

The recovered parse tree with the PR is:
```
import java.util.ArrayList;
import java.util.List;
public class SimpleJavaClass {
  public SimpleJavaClass() {
  }
  public static void main(String[] args) {
    Object unknown;
    {
      List things;
      if ((unknown instanceof List things))
          <CompleteOnType:things.is> Object;
    }
  }
}
```

@srikanth-sankaran Could you, please, review.

snjeza commented:
The completion with the PR

![pattern2](https://github.com/eclipse-jdt/eclipse.jdt.core/assets/388609/09503420-d462-402e-9843-ec268505a30a)

snjeza commented:
I have updated the PR. It works similar to CA in
```
import java.util.ArrayList;
import java.util.List;
public class SimpleJavaClass2 {
  public static void main(String[] args) {
    Object unknown = new ArrayList<>();
    List list = new ArrayList<>();
    if (unknown instanceof List things) {
      list.is|
      Object nothing = null;
      list.is
    }
    
  }
}
```
I have also added some tests.

srikanth-sankaran commented:
I apologize - I don't have the time to do justice to this task - I have been super busy with redesigning and reimplementing ECJ support for Pattern matching, Sealed classes, Switch expressions and Enhanced switch statements (all delivered already) and Records (almost done)

Later this year, I plan to comprehensively look at code assist issues - all the way from lambdas to Java 24 features. If you think this is a show stopper for using Eclipse - (I have doubts but I could be wrong) - I am happy to take myself out of the path to integrating this.

stephan-herrmann commented:
@rgrunber for the future, could you please ensure that every commit merged to master has a reference to issue and/or pull request? Otherwise looking at a commit in eclipse does not give a clue where background information can be found.

See also https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations
